### PR TITLE
Add retry to test connections

### LINKED ISSUE #22299: [Bug]: JDBCIO Write freeze at getConnection() in WriteFn
### What happened?

Re-raise BEAM-9629 as this seems not fully resolved.

Affect versions: at least >= 2.25, likely >=2.18 as in BEAM-9629.

log using 2.35.0
```
Operation ongoing in step Write Blacklisted Posts/ParDo(Write) for at least 45m00s without outputting or completing in state finish at ...
at org.apache.commons.pool2.impl.LinkedBlockingDeque.takeFirst(LinkedBlockingDeque.java:581) at 
org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:437) at 
org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:354) at 
org.apache.commons.dbcp2.PoolingDataSource.getConnection(PoolingDataSource.java:134) at 
org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:734) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.getConnection(JdbcIO.java:2029) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.executeBatch(JdbcIO.java:2087) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.finishBundle(JdbcIO.java:2048) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn$DoFnInvoker.invokeFinishBundle(Unknown Source)
```

log using 2.25.0:
```
warnings:
Operation ongoing in step Write Blacklisted Posts/ParDo(Write) for at least 45m00s without outputting or completing in state finish at ... at
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175) at
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039) at 
org.apache.commons.pool2.impl.LinkedBlockingDeque.takeFirst(LinkedBlockingDeque.java:581) at 
org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:437) at 
org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:354) at 
org.apache.commons.dbcp2.PoolingDataSource.getConnection(PoolingDataSource.java:134) at 
org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:734) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.getConnection(JdbcIO.java:2029) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.executeBatch(JdbcIO.java:2087) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn.finishBundle(JdbcIO.java:2048) at 
org.apache.beam.sdk.io.jdbc.JdbcIO$WriteFn$DoFnInvoker.invokeFinishBundle(Unknown Source)
```

### Issue Priority

Priority: 1

### Issue Component

Component: io-java-jdbc

### PR DESCRIPTION
**Pull request by johnjcasey**
fixes #22299
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by johnjcasey** [comment]
run beam_performanceTests_jdbc

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
Run Java JdbcIO Performance Test

**Comment by johnjcasey** [comment]
r: @Abacn

**Comment by Abacn** [comment]
I see,the cause is that the test did not retry create table. Can we utilize the backoff pattern here? https://github.com/apache/beam/blob/840d66a15020b5cf19c27784cb88c31b07979532/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L2345

**Comment by Abacn** [comment]
Run Java JdbcIO Performance Test

**Comment by Abacn** [comment]
Seems like org.postgresql.util.PSQLException is not caught by the catch (SQLException) https://ci-beam.apache.org/job/beam_PerformanceTests_JDBC/7101/console however PSQLException extends SQLException. Not sure why

**Comment by Abacn** [comment]
The try-with-resource statement does not catch the exception thrown in the parenthesis, I think. Replace to the simple try { getConnection } would work.

**Comment by johnjcasey** [comment]
it appears that it should catch the exception: https://stackoverflow.com/questions/15895227/try-with-resources-must-i-throw-or-catch-the-close-methods-exceptions

File: sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/DatabaseTestHelper.java [hunk 1]
```diff
@@ -100,13 +100,21 @@ public static void createTable(
...
+        connection = dataSource.getConnection();
         try (Statement statement = connection.createStatement()) {
           statement.execute(String.format("create table %s (%s)", tableName, fieldsList));
           return;
         }
       } catch (SQLException e) {
         exception = e;
+      } finally {
+        if (connection!=null){
+          connection.close();
```
Comment (Abacn) [inline sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/DatabaseTestHelper.java:116]: Probably it is closing the connecting caused issue. dataSource owns a connection pool and manage the connections. Closing it pre-maturely may have caused racing conditions.

Comment (johnjcasey) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/DatabaseTestHelper.java:116]: it looks like it is happening in .getConnection based on the stacktrace

**Comment by johnjcasey** [comment]
Run Java Precommit

**Comment by johnjcasey** [comment]
Run Java PreCommit

**Comment by johnjcasey** [comment]
@Abacn can you take another look at this?

**Review by Abacn — APPROVED** [review]
LGTM

**Pull request merged**